L44-python-rich-status.md

```diff
@@ -0,0 +1,474 @@
+    """
+```
+
+### Server side API
+
+gRPC Python team encountered lots of disagreements about how this part should be implemented. Available options are listed in the `Rationale` section. Here is our final consensus as a team, but feel free to comment about other options.
+
+During the discussion we come up with several criteria for server-side API:
+1. Shouldn't pass `ServicerContext` around and let extension package to mutate it.
+2. Should minimize the changes for our main package.
+3. Should allow the status code/message validation at some point.
+4. Should reserve the extensibility for future updates.
+5. Should not confuse developers about the priority between the new API and old APIs.
+6. Should be Pythonic.
+
+As a result, we finally decided to add a new API named `abort_with_status` that accepts the new interface `grpc.Status`. Unlike `set_code`/`set_details` can be called multiple times, the abortion mechanism ensured the code, details, and metadata of the would be set as the final status of `grpc.Call`.
+
+```Python
+def abort_with_status(self, status):
+    """Raises an exception to terminate the RPC with a non-OK status.
+
+    The status passed as argument will supercede any existing status code,
+    status message and trailing metadata.
+
+    This is an EXPERIMENTAL API.
+
+    Args:
+      status: A grpc.Status object. The status code in it must not be
+        StatusCode.OK.
+
+    Raises:
+      Exception: An exception is always raised to signal the abortion the
+        RPC to the gRPC runtime.
+    """
+```
+
+### Rich Status Usage
+
+Besides the change to status API, there should be a new gRPC Python extension package that depends on ProtoBuf named `grpcio-status`. The new package should provide convenient functions to help developers transform from ProtoBuf instance to gRPC status. The usage should look like:
+
+```Python
+# Client side
+from grpc_status import rpc_status
+
+stub = ...Stub(channel)
+
+try:
+    resp = stub.AMethodHandler(...)
+except grpc.RpcError as rpc_error:
+    rich_status = rpc_status.from_rpc_error(rpc_error)
+    # `rich_status` here is a ProtoBuf instance of `google.rpc.status.Status` proto message
+```
+```Python
+# Server side
+from grpc_status import rpc_status
+from google.protobuf import any_pb2
+
+def ...Servicer(...):
+    def ARPCCall(request, context):
+        ...
+        detail = any_pb2.Any()
+        detail.Pack(
+            rpc_status.error_details_pb2.DebugInfo(
+                stack_entries=traceback.format_stack(),
+                detail="Can't recognize this argument",
+            )
+        )
+        rich_status = grpc_status.status_pb2.Status(
+            code=grpc_status.code_pb2.INVALID_ARGUMENT,
+            message='API call quota depleted',
+            details=[detail]
+        )
+        context.abort_with_status(rpc_status.to_status(rich_status))
+        # The method handler will abort
+```
+
+### Optional: Client side API
```

Comment:
Delete this section and leave to a future gRFC?

Reply:
Removed.